src/Biblio.ts

```diff
@@ -280,9 +282,49 @@ export interface BiblioEntryBase {
   referencingIds: string[];
 }
 
+export type Type =
+  | {
+      kind: 'opaque';
+      type: string;
+    }
+  | {
+      kind: 'unused';
+    }
+  | {
+      kind: 'completion';
+      completionType: 'abrupt';
+    }
+  | {
+      kind: 'completion';
+      typeOfValueIfNormal: Type | null;
+      completionType: 'normal' | 'mixed';
+    }
+  | {
+      kind: 'list';
+      elements: Type | null;
+    }
+  | {
+      kind: 'union';
+      types: Exclude<Type, { kind: 'union' }>[];
```

Comment:
Can we exclude `{ kind: 'completion' }` here too?

Reply:
It's not totally unreasonable to have a parameter which is either a completion record or null, so I'm not going to do this right now.